sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/EventGridPublisherAsyncClient.java

```diff
@@ -57,26 +76,130 @@
+                hmac.doFinal(unsignedSas.getBytes(charset))),
+                charset);
+
+            String encodedSignature = URLEncoder.encode(signature, charset.name());
+
+            return String.format("%s&%s=%s", unsignedSas, signKey, encodedSignature);
+
+        } catch (NoSuchAlgorithmException | UnsupportedEncodingException | InvalidKeyException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException(e));
+        }
     }
 
     /**
-     * Publishes the given EventGrid events to the set topic or domain.
-     * @param events the EventGrid events to publish.
+     * Publishes the given events to the set topic or domain.
+     * @param events the events to publish.
      *
      * @return A {@link Mono} that completes when the events are sent to the service.
      * @throws NullPointerException if events is {@code null}.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
-    public Mono<Void> sendEventGridEvents(Iterable<EventGridEvent> events) {
-        return withContext(context -> sendEventGridEvents(events, context));
+    public Mono<Void> sendEvents(Iterable<T> events) {
+        return withContext(context -> sendEvents(events, context));
+    }
+
+    Mono<Void> sendEvents(Iterable<T> events, Context context) {
+        if(this.eventClass == CloudEvent.class) {
+            List<CloudEvent> eventList = new ArrayList<>();
+            events.forEach(event -> eventList.add((CloudEvent) event));
+            return this.sendCloudEvents(eventList, context);
+        } else if (this.eventClass == EventGridEvent.class) {
+            List<EventGridEvent> eventList = new ArrayList<>();
+            events.forEach(event -> eventList.add((EventGridEvent) event));
+            return this.sendEventGridEvents(eventList, context);
+        } else {
+            List<Object> eventList = new ArrayList<>();
+            events.forEach(eventList::add);
+            return this.sendCustomEvents(eventList, context);
+        }
+    }
+
+    /**
+     * Publishes the given events to the set topic or domain and gives the response issued by EventGrid.
+     * @param events the events to publish.
+     * @param context the context to use along the pipeline.
+     *
+     * @return the response from the EventGrid service.
+     * @throws NullPointerException if events is {@code null}.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<Void>> sendEventsWithResponse(Iterable<T> events, Context context) {
+        if(this.eventClass == CloudEvent.class) {
+            List<CloudEvent> eventList = new ArrayList<>();
+            events.forEach(event -> eventList.add((CloudEvent) event));
+            return this.sendCloudEventsWithResponse(eventList, context);
+        } else if (this.eventClass == EventGridEvent.class) {
+            List<EventGridEvent> eventList = new ArrayList<>();
+            events.forEach(event -> eventList.add((EventGridEvent) event));
```

Comment:
We are iterating over the events twice. It's happening once here and then again in `sendEventGridEventsWithResponse` method when converting from public EventGridEvent to internal EventGridEvent. We should avoid doing double iteration.

Reply:
I changed to use cast `(Iterable<EventGridEvent>) events` and @SuppressWarnings("unchecked") to avoid the extran iteration